### PR TITLE
Add notify-suppliers-with-signed-framework-agreement.py

### PR DESCRIPTION
**Pull request by lfdebrux**
Ticket: https://trello.com/c/KZN6ESr7/713-1-send-dos5-comms-that-framework-is-live

We want to be able to send an email to all suppliers who have signed their framework agreement when a framework goes live. This commit adds a script that sends a given Notify email template to all suppliers with a signed framework agreement.

It was almost entirely copied from `scripts/framework-applications/remind-suppliers-to-sign-framework-agreement.py`, and then modified to have the personalisations needed for Notify template "50b Services are live (DOS 5)" (98a9754d-9310-436a-9e7a-cc7504ecd92d).

I will try and add tests before we need to use it, but reviewers are encouraged to try it out with `--dry-run` :sweat_smile:.

### CODE REVIEW

File: scripts/framework-applications/notify-suppliers-with-signed-framework-agreement.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    """
+    return [supplier["supplierId"] for supplier in
+            api_client.find_framework_suppliers_iter(framework_slug, agreement_returned=True, with_declarations=False)
+            if supplier["onFramework"]]
+
+
+def get_email_addresses_for_supplier(api_client: DataAPIClient, supplier_id: int) -> List[str]:
+    """Get the email addresses for each user belonging to `supplier_id`"""
+    supplier_users = api_client.find_users_iter(supplier_id=supplier_id, personal_data_removed=False)
+    return [user["emailAddress"] for user in supplier_users if user["active"]]
```
Comment (bjgill) [inline scripts/framework-applications/notify-suppliers-with-signed-framework-agreement.py:66]: Nit: I think there's a helper that does this?

File: scripts/framework-applications/notify-suppliers-with-signed-framework-agreement.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+    Raises ValueError if the previous framework does not exist or cannot be
+    found.
+    """
+    # not convinced this is the best way to do this
+    framework_family = framework["family"]
+    frameworks_in_family = (fw for fw in api_client.find_frameworks()["frameworks"] if fw["family"] == framework_family)
+    if not frameworks_in_family:
+        raise ValueError(f"could not find any frameworks for family '{framework_family}'")
+    if len(frameworks_in_family) < 2:
```
Comment (bjgill) [inline scripts/framework-applications/notify-suppliers-with-signed-framework-agreement.py:80]: `frameworks_in_family` is a generator. Does `len` work on it?

File: scripts/framework-applications/notify-suppliers-with-signed-framework-agreement.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+
+def get_email_addresses_for_supplier(api_client: DataAPIClient, supplier_id: int) -> List[str]:
+    """Get the email addresses for each user belonging to `supplier_id`"""
+    supplier_users = api_client.find_users_iter(supplier_id=supplier_id, personal_data_removed=False)
+    return [user["emailAddress"] for user in supplier_users if user["active"]]
+
+
+def get_previous_framework(framework: dict) -> dict:
+    """Find the framework iteration that comes before `framework`
```
Comment (bjgill) [inline scripts/framework-applications/notify-suppliers-with-signed-framework-agreement.py:70]: Is this true? It looks to me like it's finding the second newest framework iteration in this family. Usually, that will be the same thing, but not necessarily.

Given the short amount of time we've got, I'm not sure it's worth working out how to do this properly, but some documentation of the pitfalls might be handy.

File: scripts/framework-applications/notify-suppliers-with-signed-framework-agreement.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+        )
+        if not DRY_RUN:
+            mail_client.send_email(
+                to_email_address=email,
+                template_name_or_id=NOTIFY_TEMPLATE_ID,
+                personalisation={
+                    "framework_name": framework["name"],
+                    "previous_framework_name": previous_framework["name"],
+                }
+            )
```
Comment (bjgill) [inline scripts/framework-applications/notify-suppliers-with-signed-framework-agreement.py:131]: I don't think we need to genericise it here, but this pattern of how to send emails to a bunch of users is starting to seem quite familiar.

**Review by bjgill — COMMENTED** [review]
A couple of questions and suggestions, but it looks broadly sound.

File: scripts/framework-applications/notify-suppliers-with-signed-framework-agreement.py (hunk 2 above)
Comment (lfdebrux) [reply scripts/framework-applications/notify-suppliers-with-signed-framework-agreement.py:80]: Excellent catch

File: scripts/framework-applications/notify-suppliers-with-signed-framework-agreement.py (hunk 3 above)
Comment (lfdebrux) [reply scripts/framework-applications/notify-suppliers-with-signed-framework-agreement.py:70]: +1

File: scripts/framework-applications/notify-suppliers-with-signed-framework-agreement.py (hunk 4 above)
Comment (lfdebrux) [reply scripts/framework-applications/notify-suppliers-with-signed-framework-agreement.py:131]: Yes, I've been thinking about that too

File: scripts/framework-applications/notify-suppliers-with-signed-framework-agreement.py (hunk 1 above)
Comment (lfdebrux) [reply scripts/framework-applications/notify-suppliers-with-signed-framework-agreement.py:72]: Probably; don't really want to dig it out right now though. This repo has a discoverability problem 🙄

File: scripts/framework-applications/notify-suppliers-with-signed-framework-agreement.py (hunk 2 above)
Comment (lfdebrux) [reply scripts/framework-applications/notify-suppliers-with-signed-framework-agreement.py:91]: Fixed

File: scripts/framework-applications/notify-suppliers-with-signed-framework-agreement.py (hunk 3 above)
Comment (lfdebrux) [reply scripts/framework-applications/notify-suppliers-with-signed-framework-agreement.py:76]: Added some notes

**Review by bjgill — APPROVED** [review]

**Pull request merged**